### PR TITLE
[4.0] Field simple color as Custom Element

### PR DESCRIPTION
**Pull request by dgrammatiko**
Pull Request for Issue # .

### Summary of Changes
- Total rewrite of the js/scss part
- dead easy HTML markup
- Accessible!
- No dependencies

### Testing Instructions
Apply the patch and check in admin the template cassiopeia

### Expected result
screen shot 2018-01-03 at 22 11 59

### Documentation Changes Required

  Kudos to @brianteeman and Christiane Maier-Stadtherr for their help on the accessibility!

### CODE REVIEW

File: layouts/joomla/form/field/color/simple.php [hunk 1]
```diff
@@ -45,25 +46,23 @@
...
-echo $onchange; ?><?php echo $autofocus; ?> style="visibility:hidden;width:22px;height:1px">
-	<?php foreach ($colors as $i => $c) : ?>
-		<option<?php echo ($c == $color ? ' selected="selected"' : ''); ?>><?php echo $c; ?></option>
-		<?php if (($i + 1) % $split == 0) : ?>
-			<option>-</option>
-		<?php endif; ?>
-	<?php endforeach; ?>
-</select>
+
+<joomla-field-simple-color text-select="Select a color" text-color="Color with hexadecimal value of " text-close="Close">
```
Comment (brianteeman) [inline layouts/joomla/form/field/color/simple.php]: The text should be in en-GB so colour

**Comment by brianteeman** [comment]
Why is it clss and not class?

**Comment by dgrammatiko** [comment]
Class is a reserved word in js we can’t use it to assign variables

File: layouts/joomla/form/field/color/simple.php (hunk 1 above)
Comment (dgrammatiko) [reply layouts/joomla/form/field/color/simple.php]: True I didn’t know if it had to be the en-GB or the lib etc

**Comment by brianteeman** [comment]
I knew there would be a good reason - thanks

**Comment by dgrammatiko** [comment]
So we do have the word close, will do that

**Comment by brianteeman** [comment]
For users of assistive technologies it is obvious that the first white box is for transparent and not for white
For the rest of us they both look white
perhaps the transparent one should be with a red line through it?

**Comment by dgrammatiko** [comment]
@brianteeman  that's the css part that needs some improvement, but I guess since all these fields will be styled properly for the new template this can be postponed. Right now the css is just an improved (?) version of the old one I guess we can do better than that. Maybe @ciar4n has something good in his mind for this

File: templates/cassiopeia/templateDetails.xml [hunk 2]
```diff
@@ -79,6 +79,13 @@
 					<option value="0">TPL_CASSIOPEIA_STATIC</option>
 				</field>
 
+				<field
```
Comment (brianteeman) [inline templates/cassiopeia/templateDetails.xml:82]: is this just for testing?
it should be Name, type, label
Label should be translatable

Comment (dgrammatiko) [reply templates/cassiopeia/templateDetails.xml:82]: Yes, the field doesn't actually connects somehow with some logic...

**Comment by C-Lodder** [comment]
Havent looked at the code yet but hope you used the dropdown code from tk

**Comment by dgrammatiko** [comment]
@C-Lodder there is no dropdown

**Comment by brianteeman** [comment]
Dont know where the color swatch came from but
1. why are we converting hex to rgb?
2. as mentioned previously wouldnt using the "real name" for assistive technology be better than reading out the hex value -" limegreen colour" is surely better than "Colour with hexadecimal value of 3, 2, C, D, 3, 2"

**Comment by dgrammatiko** [comment]
> why are we converting hex to rgb?

Well I have no clue, that part was there and I just kept it. For sure we can throw it away

About the second one: what difference will it make the `limegreen` from a hexadecimal colour code to some visual impaired? I have no answer here but we can make it work either ways (at least for the known colours)

**Comment by brianteeman** [comment]
The Colorado pallets is fixed in the library so finding their real names will be easy.

I would assume that saying  lime green is more useful and meaningful than the hex (or RGB) string

**Comment by brianteeman** [comment]
Dam autocorrect. Colorado should have been color

**Comment by dgrammatiko** [comment]
How about the following logic:
if the hex is a known color (name exists either in the basic or extended names of https://www.w3.org/TR/2017/CR-css-color-3-20171205/) then use the name else fall back to announcing the hex code

**Comment by C-Lodder** [comment]
The colour picker can use the pretty much the same code as the dropdown. Better yet, it can use it, period. Will explain on skype later

**Comment by dgrammatiko** [comment]
What exactly will a dropdown solve here?

**Comment by ciar4n** [comment]
A nice touch here would be if the available colors could be defined within the xml.

**Comment by dgrammatiko** [comment]
@ciar4n I think you can define them, that was the reason I didn't do in the first place only the named colours...

**Comment by ciar4n** [comment]
I'd be against using named colours. A user could source colours externally, copy/pasted from photoshop, sketch, a colour palette etc. And these will always be in HEX. 

For the same reason I'd be against even converting HEX to a named colour (if exists) as a user may wish to copy and paste a colour from Joomla to an external source. Becomes a pain unless you can roll off the HEX of LavenderBlush from the top of their head :smile:

**Comment by dgrammatiko** [comment]
@ciar4n we re not considering changing the actual value, only the sr text and only if the color has a known named color. So data will always be hex

**Comment by brianteeman** [comment]
yes the colors can be defined although I don't know how the colors used are either from the xml or it falls back to a standard set. 

See https://github.com/joomla/joomla-cms/blob/8e1937f7137ea37d6740dae5c55be72fa1ffbfd1/libraries/src/Form/Field/ColorField.php#L235

**Comment by ciar4n** [comment]
```
<field
	name="blabla"
	type="color"
	control="simple"
	colors="#000000, #FFFFFF, #AAAAAA"
	default=""
	label="color"
/>
```

It seems if no colors are defined in the xml then the standard set is used. The above looks likes...

image

Very neat 👍 

On a somewhat related note, we could do with a place to list topics that need to be documented for J4.

**Comment by dgrammatiko** [comment]
@ciar4n once we finish converting the fields to custom elements I will propose a neat documentation for all these, with snippets, live code examples etc, similar to the custom elements...

**Comment by dgrammatiko** [comment]
@brianteeman if you got some spare time please check this one, now will amend the colour name for the known colors (according to the W3C extended colors list)
It looks like this:
screen shot 2018-01-21 at 14 21 53

**Comment by brianteeman** [comment]
seems ok

**Comment by anuragteapot** [comment]
@dgt41 please resolve conflicts .

**Comment by dgrammatiko** [comment]
@Anu1601CS should be ok now

File: build/webcomponents/scss/field-simple-color/field-simple-color.scss [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+  button {
+    width: 20px;
+    height: 20px;
+    border: solid 1px #ccc;
+    vertical-align: middle;
+    border-radius: 3px;
+    overflow: hidden;
+    background: none;
+    -webkit-appearence: none;
```
Comment (Hils) [inline build/webcomponents/scss/field-simple-color/field-simple-color.scss:27]: appearance, not appearence

**Comment by anuragteapot** [comment]
I have tested this item :white_check_mark: successfully on 80c0e8c60c30054b799e00fe814c3b5d041816df<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19281">issues.joomla.org/tracker/joomla-cms/19281</a>.</sub>

**Comment by anuragteapot** [comment]
@astridx Please test this.

**Comment by astridx** [comment]
I have tested this item :white_check_mark: successfully on 80c0e8c60c30054b799e00fe814c3b5d041816df<br /><br />I tested this successful:

I was able to  use a default value like <tt>#222222;</tt> 
I could use the advanced mode <tt>control="advanced"</tt> and the simple mode.
I could use <tt>colors="brown, white"</tt> or <tt>colors="#AAAAAA, #000000, #FFFFFF"</tt> in simple mode.
I could use a description.

The color picker is not used in the template for changing any template colors up to now, right?
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19281">issues.joomla.org/tracker/joomla-cms/19281</a>.</sub>

**Comment by anuragteapot** [comment]
@astridx Thanks 

<code>The color picker is not used in the template for changing any template colors up to now, right?</code>
I think yes @dgt41

**Comment by ghost** [comment]
Ready to Commit after two successful tests.

**Comment by dgrammatiko** [comment]
> The color picker is not used in the template for changing any template colors up to now, right?

Yes that filed was just placed in the template for testing purposes

**Comment by laoneo** [comment]
Getting `Argument 1 passed to Joomla\CMS\HTML\HTMLHelper::webcomponent() must be of the type string, array given `.

**Comment by dgrammatiko** [comment]
@laoneo yup, these pro need an update, we recently changed the JHtml::webcomponent() to accept the file path similar to stylesheet and script. Will do that in a bit for all 
PS ok I was too slow...

**Comment by laoneo** [comment]
Can you fix conflicts here?

**Comment by laoneo** [comment]
@dgt41 fixed it in https://github.com/joomla/joomla-cms/pull/19281/commits/8a76be81f6911a0c3cc39f7df97485481a28dcd1

**Comment by tassosm** [comment]
It doesn't work anymore. The selected color doesn't get saved as the select element is not updated.

**Comment by dgrammatiko** [comment]
@tassosm thanks for testing, here is the fix: https://github.com/joomla/joomla-cms/pull/20146
I think we consecrated our efforts to meet A11Y and we missed that one

**Pull request merged**